Vwitter/SpotifyAPI/SpotifyAPIManager.m

```diff
@@ -0,0 +1,212 @@
+//
+//  SpotifyAPIManager.m
+//  Vwitter
+//
+//  Created by Christina Li on 7/28/22.
+//
+
+#import "SpotifyAPIManager.h"
+#import "AppDelegate.h"
+#import "VWHelpers.h"
+#import "SpotifyTrack.h"
+
+static NSString * const SpotifyClientID = @"e4185723643e4db9bcf48af28e078cff";
+static NSString * const SpotifyRedirectURLString = @"vwitter://callback/";
+
+@interface SpotifyAPIManager ()
+
+@property (strong, nonatomic, nullable) NSString *accessToken;
+@property (strong, nonatomic, nullable) NSNumber *accessTokenExpirationTimestampSeconds;
+
+@end
+
+@implementation SpotifyAPIManager
+
+static NSString *const kExpirationKey = @"spotify_expires_timestamp";
+
++ (instancetype)shared {
+    static SpotifyAPIManager *sharedManager = nil;
+    static dispatch_once_t onceToken;
+    dispatch_once(&onceToken, ^{
+        sharedManager = [[self alloc] init];
+    });
+    return sharedManager;
+}
+
+- (void)didAuthorizeWithSpotify:(NSString *)accessToken expiresInSeconds:(NSNumber *)expiresInSeconds {
+    self.accessToken = accessToken;
+    self.accessTokenExpirationTimestampSeconds = @(NSDate.date.timeIntervalSince1970 + expiresInSeconds.doubleValue);
+    self.appRemote.connectionParameters.accessToken = accessToken;
+}
+
+- (BOOL)hasValidSpotifyAuthorization {
+    NSTimeInterval timestampSeconds = NSDate.date.timeIntervalSince1970;
+    return (self.accessToken != nil
+            && self.accessTokenExpirationTimestampSeconds != nil
+            && self.accessTokenExpirationTimestampSeconds.doubleValue > timestampSeconds);
+}
+
+- (void)setupAppRemote {
+    /*
+     Scopes let you specify exactly what types of data your application wants to
+     access, and the set of scopes you pass in your call determines what access
+     permissions the user is asked to grant.
+     For more information, see https://developer.spotify.com/web-api/using-scopes/.
+     */
+    SPTScope scope = SPTUserLibraryReadScope | SPTPlaylistReadPrivateScope;
+
+    /*
+     Start the authorization process. This requires user input.
+     */
+    if (@available(iOS 11, *)) {
+        // Use this on iOS 11 and above to take advantage of SFAuthenticationSession
+        [self.sessionManager initiateSessionWithScope:scope options:SPTDefaultAuthorizationOption];
+        NSLog(@"authorized spotify session");
+    } else {
+        // Use this on iOS versions < 11 to use SFSafariViewController
+        NSLog(@"iOS version too old");
```

Comment:
should we do an early return here? Great job printing the error, this makes it much easier to debug problems